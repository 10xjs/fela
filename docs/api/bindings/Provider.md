# Provider

The Provider Component has been renamed to [RendererProvider](./RendererProvider).