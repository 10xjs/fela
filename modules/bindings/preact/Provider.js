/* @flow weak */
import { Component } from 'preact'
import render from '../../dom/render'

export default class Provider extends Component {
  getChildContext() {
    return { renderer: this.props.renderer }
  }

  componentDidMount() {
    const { mountNode, renderer } = this.props

    if (mountNode) {
      render(renderer, mountNode)
    }
  }

  render({ children }) {
    return children
  }
}
