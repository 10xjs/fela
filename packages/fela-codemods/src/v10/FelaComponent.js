import getImportName from '../utils/getImportName'

const pkg = 'react-fela'
const component = 'FelaComponent'

export default function transformer(file, api) {
  const j = api.jscodeshift

  const importName = getImportName(j, file.source, pkg, component)

  // only transform if FelaComponent is imported from react-fela
  if (importName) {
    return j(file.source)
      .find(j.JSXElement)
      .forEach(path => {
        if (path.node.openingElement.name.name === importName) {
          const styleProp = path.node.openingElement.attributes.find(
            prop => prop.name.name === 'style'
          )

          const ruleProp = path.node.openingElement.attributes.find(
            prop => prop.name.name === 'rule'
          )

          const renderProp = path.node.openingElement.attributes.find(
            prop => prop.name.name === 'render'
          )

          // handle render/as transformation to children/as
          if (renderProp) {
            if (
              renderProp.value.type === 'Literal' ||
              (renderProp.value.type === 'JSXExpressionContainer' &&
                renderProp.value.expression.type === 'Literal')
            ) {
              renderProp.name.name = 'as'
            } else {
              j(path).replaceWith(
                j.jsxElement(
                  j.jsxOpeningElement(
                    j.jsxIdentifier(importName),
                    path.node.openingElement.attributes.filter(
                      prop => prop !== renderProp
                    ),
                    false
                  ),
                  j.jsxClosingElement(j.jsxIdentifier(importName)),
                  [renderProp.value],
                  false
                )
              )
            }
          }

          // replace inline style expression from theme
          if (styleProp) {
            j(styleProp)
              .find(j.ArrowFunctionExpression)
              .forEach(stylePath => {
                const param = stylePath.node.params[0]

                if (param.type === 'Identifier') {
                  j(stylePath).replaceWith(
                    j.arrowFunctionExpression(
                      [
                        j.objectPattern([
                          j.objectProperty(param, param, false, true),
                        ]),
                      ],
                      stylePath.node.body
                    )
                  )

                  // make property truly shorthand
                  stylePath.node.params[0].properties[0].shorthand = true
                }
              })
          }

          // rename rule to style
          if (ruleProp) {
            ruleProp.name.name = 'style'
          }
        }
      })
      .toSource()
  }

  return file.source
}
