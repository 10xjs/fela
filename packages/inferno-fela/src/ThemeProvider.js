/* @flow */
import { Component } from 'inferno'
import { createElement } from 'inferno-create-element'
import { ThemeProviderFactory } from 'fela-bindings'

import { ThemeContext } from './context'

export default ThemeProviderFactory(
  Component,
  ThemeContext,
  createElement,
  children => children,
  {
    defaultProps: {
      overwrite: false,
    },
  }
)
