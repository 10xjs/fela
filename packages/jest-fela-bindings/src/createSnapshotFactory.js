import os from 'os'
import { format } from 'prettier'
import HTMLtoJSX from 'htmltojsx'

import { renderToString } from 'fela-tools'

import type { DOMRenderer } from '../../../flowtypes/DOMRenderer'

function formatCSS(css) {
  return format(css, { parser: 'css', useTabs: false, tabWidth: 2 })
}

function formatHTML(html) {
  const converter = new HTMLtoJSX({
    createClass: false,
  })

  const jsx = converter.convert(html)
  return format(jsx).replace(/[\\"]/g, '')
}

export default function createSnapshotFactory(
  createElement: Function,
  render: Function,
  defaultRenderer: Function,
  defaultProvider: Function,
  defaultThemeProvider: Function
): Function {
  return function createSnapshot(
    component: any,
    theme: Object = {},
    renderer: DOMRenderer = defaultRenderer,
    Provider: Function = defaultProvider,
    ThemeProvider: Function = defaultThemeProvider
  ) {
    const div = document.createElement('div')

    // reset renderer to have a clean setup
    renderer.clear()

    render(
      createElement(
        Provider,
        { renderer },
        createElement(ThemeProvider, { theme }, component)
      ),
      div
    )

    return (
      formatCSS(renderToString(renderer)) +
      os.EOL +
      os.EOL +
      formatHTML(div.innerHTML)
    )
  }
}
