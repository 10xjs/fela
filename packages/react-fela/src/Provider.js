/* @flow */
import { Component } from 'react'
import { ProviderFactory } from 'fela-bindings'
import PropTypes from 'prop-types'

export default ProviderFactory(Component, children => children, {
  propTypes: {
    renderer: PropTypes.object.isRequired,
    rehydrate: PropTypes.bool.isRequired,
  },
  childContextTypes: {
    renderer: PropTypes.object,
  },
  defaultProps: {
    rehydrate: true,
  },
})
