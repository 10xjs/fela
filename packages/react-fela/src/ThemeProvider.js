/* @flow */
import { Component } from 'react'
import { ThemeProviderFactory } from 'fela-bindings'
import PropTypes from 'prop-types'

export default ThemeProviderFactory(Component, children => children, {
  propTypes: {
    theme: PropTypes.object.isRequired,
    overwrite: PropTypes.bool
  },
  childContextTypes: {
    theme: PropTypes.object
  },
  contextTypes: {
    theme: PropTypes.object
  },
  defaultProps: {
    overwrite: false
  }
})
