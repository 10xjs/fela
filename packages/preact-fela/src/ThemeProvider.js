/* @flow */
// $FlowFixMe
import { Component, h as createElement } from 'preact'
import { ThemeProviderFactory } from 'fela-bindings'

import { ThemeContext } from './context'

export default ThemeProviderFactory(
  Component,
  ThemeContext,
  createElement,
  ([children]) => children,
  {
    defaultProps: {
      overwrite: false,
    },
  }
)
