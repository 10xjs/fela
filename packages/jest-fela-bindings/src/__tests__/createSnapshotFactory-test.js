import React, { createElement } from 'react'
import { render } from 'react-dom'

import createSnapshotFactory from '../createSnapshotFactory'

import FelaComponent from '../../../react-fela/src/FelaComponent'
import Provider from '../../../react-fela/src/Provider'
import ThemeProvider from '../../../react-fela/src/ThemeProvider'
import createRenderer from '../../../fela/src/createRenderer'
import plugins from '../../../fela-preset-web/src/index'

const createSnapshot = createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)

describe('Creating Snapshots with Fela', () => {
  it('should return formatted html and css', () => {
    const style = {
      display: 'flex',
      flex: '1 0 auto',
      ':hover': {
        color: 'blue',
      },
      '@media (min-width: 300px)': {
        backgroundColor: 'black',
        ':hover': {
          color: 'green',
        },
      },
    }

    expect(createSnapshot(<FelaComponent style={style} />)).toMatchSnapshot()
  })

  it('should always use a clean setup', () => {
    expect(
      createSnapshot(
        <FelaComponent style={{ color: 'red', backgroundColor: 'blue' }} />
      )
    ).toMatchSnapshot()
    expect(
      createSnapshot(
        <FelaComponent style={{ color: 'red', backgroundColor: 'green' }} />
      )
    ).toMatchSnapshot()
  })

  it('should inject the theme', () => {
    const rule = ({ theme }) => ({
      backgroundColor: theme.bgColor,
      color: 'red',
    })

    expect(
      createSnapshot(<FelaComponent rule={rule} />, { bgColor: 'blue' })
    ).toMatchSnapshot()
  })

  it('should use a custom renderer', () => {
    const renderer = createRenderer({
      plugins,
    })

    expect(
      createSnapshot(
        <FelaComponent style={{ appearance: 'none', fontSize: 12 }} />,
        {},
        renderer
      )
    ).toMatchSnapshot()
  })
})
