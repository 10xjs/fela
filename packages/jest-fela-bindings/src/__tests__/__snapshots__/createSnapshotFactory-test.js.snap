// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating Snapshots with Fela should always use a clean setup 1`] = `
".a {
  background-color: blue;
}
.b {
  color: red;
}


<div data-reactroot className=a b />;
"
`;

exports[`Creating Snapshots with Fela should always use a clean setup 2`] = `
".a {
  background-color: green;
}
.b {
  color: red;
}


<div data-reactroot className=a b />;
"
`;

exports[`Creating Snapshots with Fela should inject the theme 1`] = `
".a {
  background-color: blue;
}
.b {
  color: red;
}


<div data-reactroot className=a b />;
"
`;

exports[`Creating Snapshots with Fela should return formatted html and css 1`] = `
".d {
  display: flex;
}
.e {
  flex: 1 0 auto;
}
.a:hover {
  color: blue;
}
@media (min-width: 300px) {
  .c {
    background-color: black;
  }
  .b:hover {
    color: green;
  }
}


<div data-reactroot className=a b c d e />;
"
`;

exports[`Creating Snapshots with Fela should use a custom renderer 1`] = `
".a {
  font-size: 12px;
}
.b {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}


<div data-reactroot className=a b />;
"
`;
